src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinderProvider.cs

```diff
@@ -52,6 +52,17 @@ public BodyModelBinderProvider(IList<IInputFormatter> formatters, IHttpRequestSt
             _loggerFactory = loggerFactory;
         }
 
+        /// <summary>
+        /// Gets or sets a flag that controls whether the binder should require non-null
+        /// input values to be supplied. Defaults to <c>true</c>.
+        /// </summary>
+        /// <example>
+        /// If <c>true</c>, and an associated <see cref="IInputFormatter"/> supplies <c>null</c>
+        /// as the value to bind (for example, if an incoming POST or PUT request supplies a
+        /// zero-length body), the binder will register a validation error.
+        /// </example>
+        public bool IsBindingRequired { get; set; } = true;
```

Comment:
>  Also add an `IBindingMetadataProvider` implementation (or extend `DefaultBindingMetadataProvider`) similar to `DataMemberRequiredBindingMetadataProvider` but based on the calculated `BindingSource` and the new option.

Ignore this part. Don't know where my brain went. @rynowak reminded me this only works for properties and types. Our attributes that implement `IBindingSourceMetadata` can't be associated with a type and, as @SteveSandersonMS noted earlier, we don't get unique `ModelMetadata` for parameters -- just the type's info.

Instead, add constructors accepting `MvcOptions` parameters to `BodyModelBinderProvider` and `BodyModelBinder`. Also add a property of this type to the `InputFormatterContext` class.